File: java/L300/LeetCode0331.java
package L300;

public class LeetCode0331 {
    private int i;
    public boolean isValidSerialization(String preorder) {
        if (preorder == null || preorder.length() == 0) {
            return true;
        }
        i = 0;
        return isValidIndex(preorder) && i >= preorder.length();
    }

    private boolean isValidIndex(String tree) {
        final int N = tree.length();
        while (i < N && tree.charAt(i) == ' ') {
            ++i;
        }
        if (i >= N) {
            return false;
        }
        if (tree.charAt(i) == '#') {
            while (i < N && tree.charAt(i) != ',') {
                ++i;
            }
            ++i;
            return true;
        } else if (Character.isDigit(tree.charAt(i))) {
            while (i < N && tree.charAt(i) != ',') {
                ++i;
            }
            ++i;
            boolean flag = isValidIndex(tree);
            flag &= isValidIndex(tree);
            return flag;
        }
        return false;
    }
}

File: cpp/L300/LeetCode0331.cc
#include <string>
using std::string;

class Solution {

    int validatePreOrder(string const &preorder, int index) {
        int N = preorder.size();
        if (index >= N) {
            return -1;
        }
        if (preorder[index] == '#') {
            return index + 1 < N ? index + 2 : index + 1;
        }
        while (index < N && preorder[index] != ',') {
            ++index;
        }
        if (index == N || index + 1 == N) {
            return -1;
        }
        int lx = validatePreOrder(preorder, index + 1);
        if (lx == -1 || lx == N) {
            return -1;
        }
        return validatePreOrder(preorder, lx);
    }
public:
    bool isValidSerialization(string preorder) {
        int x = validatePreOrder(preorder, 0);
        return x == preorder.size();
    }
};